sbin/newfs/newfs.8
.\"	$NetBSD$
.\"
.\" Copyright (c) 1983, 1987, 1991, 1993, 1994
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)newfs.8	8.6 (Berkeley) 5/3/95
.\"
.Dd June 30, 2012
.Dt NEWFS 8
.Os
.Sh NAME
.Nm newfs
.Nd construct a new file system
.Sh SYNOPSIS
.Nm
.Op Fl FGINZ
.Op Fl a Ar maxcontig
.Op Fl B Ar byte-order
.Op Fl b Ar block-size
.Op Fl d Ar maxbsize
.Op Fl e Ar maxbpg
.Op Fl f Ar frag-size
.Op Fl g Ar avgfilesize
.Op Fl h Ar avgfpdir
.Op Fl i Ar bytes-per-inode
.Op Fl m Ar free-space
.Op Fl n Ar inodes
.Op Fl O Ar filesystem-format
.Op Fl o Ar optimization
.Op Fl q Ar quota
.Op Fl S Ar sector-size
.Op Fl s Ar size
.Op Fl T Ar disk-type
.Op Fl v Ar volname
.Op Fl V Ar verbose
.Ar special
.Sh DESCRIPTION
.Nm
is used to initialize and clear file systems before first use.
Before running
.Nm
the disk must be labeled using
.Xr disklabel 8 .
.Nm
builds a file system on the specified special device
basing its defaults on the information in the disk label.
Typically the defaults are reasonable, however
.Nm
has numerous options to allow the defaults to be selectively overridden.
.Pp
Options with numeric arguments may contain an optional (case-insensitive)
suffix:
.Bl -tag -width 3n -offset indent -compact
.It b
Bytes; causes no modification.
(Default)
.It k
Kilo; multiply the argument by 1024.
.It m
Mega; multiply the argument by 1048576.
.It g
Giga; multiply the argument by 1073741824.
.It t
Tera; multiply the argument by 1099511627776.
.El
.Pp
The following options define the general layout policies.
.Bl -tag -width Fl
.It Fl a Ar maxcontig
This sets the obsolete maxcontig parameter.
.It Fl B Ar byte-order
Specify the metadata byte order of the file system to be created.
Valid byte orders are
.Sq be
and
.Sq le .
If no byte order is specified, the file system is created in host
byte order.
.It Fl b Ar block-size
The block size of the file system, in bytes.
It must be a power of two.
The smallest allowable size is 4096 bytes.
The default size depends upon the size of the file system:
.Pp
.Bl -tag -width "file system size" -compact -offset indent
.It Sy "file system size"
.Ar block-size
.It \*[Lt] 20 MB
4 KB
.It \*[Lt] 1000 MB
8 KB
.It \*[Lt] 128 GB
16 KB
.It \*[Gt]= 128 GB
32 KB
.El
.It Fl d Ar maxbsize
Set the maximum extent size to
.Ar maxbsize .
.It Fl e Ar maxbpg
This indicates the maximum number of blocks any single file can
allocate out of a cylinder group before it is forced to begin
allocating blocks from another cylinder group.
The default is about one quarter of the total blocks in a cylinder group.
See
.Xr tunefs 8
for more details on how to set this option.
.It Fl F
Create a file system image in
.Ar special .
The file system size needs to be specified with
.Dq Fl s Ar size .
No attempts to use or update the disk label will be made.
.It Fl f Ar frag-size
The fragment size of the file system in bytes.
It must be a power of two ranging in value between
.Ar block-size Ns /8
and
.Ar block-size .
The optimal
.Ar block-size : Ns Ar frag-size
ratio is 8:1.
Other ratios are possible, but are not recommended,
and may produce unpredictable results.
The default size depends upon the size of the file system:
.Pp
.Bl -tag -width "file system size" -compact -offset indent
.It Sy "file system size"
.Ar frag-size
.It \*[Lt] 20 MB
0.5 KB
.It \*[Lt] 1000 MB
1 KB
.It \*[Lt] 128 GB
2 KB
.It \*[Gt]= 128 GB
4 KB
.El
.It Fl G
Treat garbage parameters as non-fatal.
Using this option may result in a file system which causes a kernel
panic and should only be used for testing.
.It Fl g Ar avgfilesize
The expected average file size for the file system.
.It Fl h Ar avgfpdir
The expected average number of files per directory on the file system.
.It Fl I
Do not require that the file system type listed in the disk label is
.Ql 4.2BSD
or
.Ql Apple UFS .
.It Fl i Ar bytes-per-inode
This specifies the density of inodes in the file system.
If fewer inodes are desired, a larger number should be used;
to create more inodes a smaller number should be given.
The default is to create an inode for every
.Pq 4 * Ar frag-size
bytes of data space:
.Pp
.Bl -tag -width "file system size" -compact -offset indent
.It Sy "file system size"
.Ar bytes-per-inode
.It \*[Lt] 20 MB
2 KB
.It \*[Lt] 1000 MB
4 KB
.It \*[Lt] 128 GB
8 KB
.It \*[Gt]= 128 GB
16 KB
.El
.It Fl m Ar free-space
The percentage of space reserved from normal users; the minimum free
space threshold.
The default value used is 5%.
See
.Xr tunefs 8
for more details on how to set this option.
.It Fl N
Causes the file system parameters to be printed out
without really creating the file system.
.It Fl n Ar inodes
This specifies the number of inodes for the filesystem.
If both
.Fl i
and
.Fl n
are specified then
.Fl n
takes precedence.
.It Fl O Ar filesystem-format
Select the filesystem-format.
.Bl -tag -width 3n -offset indent -compact
.It 0
4.3BSD; This option is primarily used to build root file systems that can be
understood by older boot ROMs.
This generates an FFSv1 file system with level 1 format.
.It 1
FFSv1; normal Fast File System, level 4 format.
Also known as
.Sq FFS ,
.Sq UFS ,
or
.Sq UFS1 .
This is the default.
.It 2
FFSv2; enhanced Fast File System, suited for more than 1 Terabyte capacity.
.\" Supports access control lists.
This is also known as
.Sq UFS2 .
.El
See
.Xr fsck_ffs 8
for more information about format levels.
.Pp
To create an LFS filesystem see
.Xr newfs_lfs 8 .
To create a Linux ext2 filesystem see
.Xr newfs_ext2fs 8 .
.It Fl o Ar optimization
Optimization preference; either
.Dq space
or
.Dq time .
The file system can either be instructed to try to minimize the time spent
allocating blocks, or to try to minimize the space fragmentation on the disk.
If the value of minfree (see above) is less than 5%,
the default is to optimize for space;
if the value of minfree is greater than or equal to 5%,
the default is to optimize for time.
See
.Xr tunefs 8
for more details on how to set this option.
.It Fl q Ar quota
enable a quota.
.Ar quota
can be one of
.Li user
or
.Li group
to enable the specified quota type.
Multiple
.Fl q
can be used to enable all types at once.
.It Fl s Ar size
The size of the file system in sectors.
An
.Sq s
suffix will be interpreted as the number of sectors (the default).
All other suffixes are interpreted as per other numeric arguments,
except that the number is converted into sectors by dividing by the
sector size (as specified by
.Fl S Ar secsize )
after suffix interpretation.
.Pp
If no
.Fl s Ar size
is specified then the filesystem size defaults to that of the partition, or,
if
.Fl F
is specified, the existing file.
.Pp
If
.Ar size
is negative the specified size is subtracted from the default size
(reserving space at the end of the partition).
.It Fl T Ar disk-type
Uses information for the specified disk from
.Pa /etc/disktab
instead of trying to get the information from the disk label.
.It Fl V Ar verbose
This controls the amount of information written to stdout:
.Bl -tag -width 3n -offset indent -compact
.It 0
No output.
.It 1
Overall size and cylinder group details.
.It 2
A progress bar (dots ending at right hand margin).
.It 3
The first few super-block backup sector numbers are displayed before the
progress bar.
.It 4
All the super-block backup sector numbers are displayed (no progress bar).
.El
The default is 3.
If
.Fl N
is specified
.Nm
stops before outputting the progress bar.
.It Fl v Ar volname
This specifies that an Apple UFS filesystem should be created
with the given volume name.
.It Fl Z
Pre-zeros the file system image created with
.Fl F .
This is necessary if the image is to be used by
.Xr vnd 4
(which doesn't support file systems with
.Sq holes ) .
.El
.Pp
The following option overrides the standard sizes for the disk geometry.
The default value is taken from the disk label.
Changing this default is useful only when using
.Nm
to build a file system whose raw image will eventually be used on a
different type of disk than the one on which it is initially created
(for example on a write-once disk).
Note that changing this value from its default will make
it impossible for
.Xr fsck_ffs 8
to find the alternative superblocks if the standard superblock is lost.
.Bl -tag -width Fl
.It Fl S Ar sector-size
The size of a sector in bytes (almost never anything but 512).
Defaults to 512.
.El
.Sh NOTES
The file system is created with
.Sq random
inode generation numbers to improve NFS security.
.Pp
The owner and group IDs of the root node of the new file system
are set to the effective UID and GID of the user initializing
the file system.
.Pp
For the
.Nm
command to succeed,
the disk label should first be updated such that the fstype field for the
partition is set to
.Ql 4.2BSD
or
.Ql Apple UFS ,
unless
.Fl F
or
.Fl I
is used.
.Pp
To create and populate a filesystem image within a file use the
.Xr makefs 8
utility.
.Pp
The partition size is found using
.Xr fstat 2 ,
not by inspecting the disk label.
The block size and fragment size will be written back to the disk label
only if the last character of
.Ar special
references the same partition as the minor device number.
.Pp
Unless
.Fl F
is specified,
.Ar special
must be a raw device.
This means that for example
.Pa wd0a
or
.Pa /dev/rwd0a
must be specified instead of
.Pa /dev/wd0a .
.Sh SEE ALSO
.Xr fstat 2 ,
.Xr disktab 5 ,
.Xr fs 5 ,
.Xr disklabel 8 ,
.Xr diskpart 8 ,
.Xr dumpfs 8 ,
.\" .Xr format 8 ,
.Xr fsck_ffs 8 ,
.\" .Xr fsirand 8 ,
.Xr makefs 8 ,
.Xr mount 8 ,
.Xr mount_mfs 8 ,
.Xr newfs_ext2fs 8 ,
.Xr newfs_lfs 8 ,
.Xr newfs_msdos 8 ,
.Xr tunefs 8
.Rs
.%A M. McKusick
.%A W. Joy
.%A S. Leffler
.%A R. Fabry
.%T A Fast File System for UNIX ,
.%J ACM Transactions on Computer Systems 2
.%V 3
.%P pp 181-197
.%D August 1984
.%O (reprinted in the BSD System Manager's Manual)
.Re
.Rs
.%A M. McKusick
.%T Enhancements to the fast filesystem to support multi-terabyte storage systems
.%J Proceedings of the BSD Conference 2003
.%P pp 79-90
.%D September 2003
.Re
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

sbin/newfs_lfs/newfs.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 1989, 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1989, 1992, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)newfs.c	8.5 (Berkeley) 5/24/95";
#else
__RCSID("$NetBSD$");
#endif
#endif /* not lint */

/*
 * newfs: friendly front end to mkfs
 */
#include <sys/param.h>
#include <sys/ucred.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/file.h>
#include <sys/mount.h>
#include <sys/sysctl.h>
#include <sys/time.h>
#include <sys/disk.h>

#include <ufs/lfs/lfs.h>

#include <err.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <paths.h>
#include <util.h>
#include "config.h"
#include "extern.h"
#include "bufcache.h"
#include "partutil.h"

#define	COMPAT			/* allow non-labeled disks */

#ifdef COMPAT
const char lmsg[] = "%s: can't read disk label; disk type must be specified";
#else
const char lmsg[] = "%s: can't read disk label";
#endif

int	Nflag = 0;		/* run without writing file system */
int	fssize;			/* file system size */
int	sectorsize;		/* bytes/sector */
int	fsize = 0;		/* fragment size */
int	bsize = 0;		/* block size */
int	ibsize = 0;		/* inode block size */
int	interleave = 0;		/* segment interleave */
int	minfree = MINFREE;	/* free space threshold */
int     minfreeseg = 0;         /* segments not counted in bfree total */
int     resvseg = 0;            /* free segments reserved for the cleaner */
u_int32_t roll_id = 0;		/* roll-forward id */
u_long	memleft;		/* virtual memory available */
caddr_t	membase;		/* start address of memory based filesystem */
#ifdef COMPAT
char	*disktype;
#endif
int	preen = 0;		/* Coexistence with fsck_lfs */

char	device[MAXPATHLEN];
char	*progname, *special;

extern long	dev_bsize;		/* device block size */

static int64_t strsuftoi64(const char *, const char *, int64_t, int64_t, int *);
static void usage(void);

/* CHUNKSIZE should be larger than MAXPHYS */
#define CHUNKSIZE (1024 * 1024)

static size_t
auto_segsize(int fd, off_t len, int version)
{
	off_t off, bw;
	time_t start, finish;
	char buf[CHUNKSIZE];
	long seeks;
	size_t final;
	int i;
	
	/* First, get sequential access bandwidth */
	time(&start);
	finish = start;
	for (off = 0; finish - start < 10; off += CHUNKSIZE) {
		if (pread(fd, buf, CHUNKSIZE, off) < 0)
			break;
		time(&finish);
	}
	/* Bandwidth = bytes / sec */
	/* printf("%ld bytes in %ld seconds\n", (long)off, (long)(finish - start)); */
	bw = off / (finish - start);

	/* Second, seek time */
	time(&start);
	finish = start; /* structure copy */
	for (seeks = 0; finish - start < 10; ) {
		off = (((double)rand()) * (btodb(len))) / ((off_t)RAND_MAX + 1);
		if (pread(fd, buf, dbtob(1), dbtob(off)) < 0)
			err(1, "pread");
		time(&finish);
		++seeks;
	}
	/* printf("%ld seeks in %ld seconds\n", (long)seeks, (long)(finish - start)); */
	/* Seek time in units/sec */
	seeks /= (finish - start);
	if (seeks == 0)
		seeks = 1;

	printf("bw = %ld B/s, seek time %ld ms (%ld seeks/s)\n",
		(long)bw, 1000/seeks, seeks);
	final = dbtob(btodb(4 * bw / seeks));
	if (version == 1) {
		for (i = 0; final; final >>= 1, i++)
			;
		final = 1 << i;
	}
	printf("using initial segment size %ld\n", (long)final);
	return final;
}

int
main(int argc, char **argv)
{
	int version, ch;
	struct disk_geom geo;
	struct dkwedge_info dkw;
	struct stat st;
	int debug, force, fsi, fso, lfs_segsize, maxpartitions;
	uint secsize = 0;
	daddr_t start;
	const char *opstring;
	int byte_sized = 0;
	int r;

	version = DFL_VERSION;		/* what version of lfs to make */

	if ((progname = strrchr(*argv, '/')) != NULL)
		++progname;
	else
		progname = *argv;

	maxpartitions = getmaxpartitions();
	if (maxpartitions > 26)
		fatal("insane maxpartitions value %d", maxpartitions);

	opstring = "AB:b:DFf:I:i:LM:m:NO:R:r:S:s:v:";

	debug = force = lfs_segsize = start = 0;
	while ((ch = getopt(argc, argv, opstring)) != -1)
		switch(ch) {
		case 'A':	/* Adaptively configure segment size */
			lfs_segsize = -1;
			break;
		case 'B':	/* LFS segment size */
		        lfs_segsize = strsuftoi64("segment size", optarg, LFS_MINSEGSIZE, INT64_MAX, NULL);
			break;
		case 'D':
			debug = 1;
			break;
		case 'F':
			force = 1;
			break;
		case 'I':
		        interleave = strsuftoi64("interleave", optarg, 0, INT64_MAX, NULL);
			break;
		case 'L':	/* Compatibility only */
			break;
		case 'M':
		  	minfreeseg = strsuftoi64("minfreeseg", optarg, 0, INT64_MAX, NULL);
			break;
		case 'N':
			Nflag++;
			break;
		case 'O':
		  	start = strsuftoi64("start", optarg, 0, INT64_MAX, NULL);
			break;
		case 'R':
		  	resvseg = strsuftoi64("resvseg", optarg, 0, INT64_MAX, NULL);
			break;
		case 'S':
		  	secsize = strsuftoi64("sector size", optarg, 1, INT64_MAX, NULL);
			if (secsize <= 0 || (secsize & (secsize - 1)))
				fatal("%s: bad sector size", optarg);
			break;
#ifdef COMPAT
		case 'T':
			disktype = optarg;
			break;  
#endif 
		case 'b':
		  	bsize = strsuftoi64("block size", optarg, LFS_MINBLOCKSIZE, INT64_MAX, NULL);
			break;
		case 'f':
		  	fsize = strsuftoi64("fragment size", optarg, LFS_MINBLOCKSIZE, INT64_MAX, NULL);
			break;
		case 'i':
		  	ibsize = strsuftoi64("inode block size", optarg, LFS_MINBLOCKSIZE, INT64_MAX, NULL);
			break;
		case 'm':
		  	minfree = strsuftoi64("free space %", optarg, 0, 99, NULL);
			break;
		case 'r':
		  	roll_id = strsuftoi64("roll-forward id", optarg, 1, UINT_MAX, NULL);
			break;
		case 's':
		        fssize = strsuftoi64("file system size", optarg, 0, INT64_MAX, &byte_sized);
			break;
		case 'v':
		        version = strsuftoi64("file system version", optarg, 1, LFS_VERSION, NULL);
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc != 2 && argc != 1)
		usage();

	/*
	 * If the -N flag isn't specified, open the output file.  If no path
	 * prefix, try /dev/r%s and then /dev/%s.
	 */
	special = argv[0];
	if (strchr(special, '/') == NULL) {
		(void)snprintf(device, sizeof(device), "%sr%s", _PATH_DEV,
		    special);
		if (stat(device, &st) == -1)
			(void)snprintf(device, sizeof(device), "%s%s",
			    _PATH_DEV, special);
		special = device;
	}
	if (!Nflag) {
		fso = open(special, O_RDWR, DEFFILEMODE);
		if (debug && fso < 0) {
			/* Create a file of the requested size. */
			fso = open(special, O_CREAT | O_RDWR, DEFFILEMODE);
			if (fso >= 0) {
				char buf[512];
				int i;
				for (i = 0; i < fssize; i++)
					write(fso, buf, sizeof(buf));
				lseek(fso, 0, SEEK_SET);
			}
		}
		if (fso < 0)
			fatal("%s: %s", special, strerror(errno));
	} else
		fso = -1;

	/* Open the input file. */
	fsi = open(special, O_RDONLY);
	if (fsi < 0)
		fatal("%s: %s", special, strerror(errno));
	if (fstat(fsi, &st) < 0)
		fatal("%s: %s", special, strerror(errno));


	if (!S_ISCHR(st.st_mode)) {
		if (!S_ISREG(st.st_mode)) {
			fatal("%s: neither a character special device "
			      "nor a regular file", special);
		}
		(void)strcpy(dkw.dkw_ptype, DKW_PTYPE_LFS); 
		if (secsize == 0)
			secsize = 512;
		dkw.dkw_size = st.st_size / secsize;
	} else {
#ifdef COMPAT
		if (disktype == NULL)
			disktype = argv[1];
#endif
		if (getdiskinfo(special, fsi, disktype, &geo, &dkw) == -1)
			errx(1, lmsg, special);

		if (dkw.dkw_size == 0)
			fatal("%s: is zero sized", argv[0]);
		if (!force && strcmp(dkw.dkw_ptype, DKW_PTYPE_LFS) != 0)
			fatal("%s: is not `%s', but `%s'", argv[0],
			    DKW_PTYPE_LFS, dkw.dkw_ptype);
	}

	if (secsize == 0)
		secsize = geo.dg_secsize;

	/* Make device block size available to low level routines */
	dev_bsize = secsize;

	/* From here on out fssize is in sectors */
	if (byte_sized) {
		fssize /= secsize;
	}

	/* If force, make the partition look like an LFS */
	if (force) {
		(void)strcpy(dkw.dkw_ptype, DKW_PTYPE_LFS); 
		if (fssize) {
			dkw.dkw_size = fssize;
		}
	} else
		if (fssize != 0 && fssize < dkw.dkw_size)
			dkw.dkw_size = fssize;

	/* Try autoconfiguring segment size, if asked to */
	if (lfs_segsize == -1) {
		if (!S_ISCHR(st.st_mode)) {
			warnx("%s is not a character special device, ignoring -A", special);
			lfs_segsize = 0;
		} else
			lfs_segsize = auto_segsize(fsi, dkw.dkw_size / secsize,
			    version);
	}

	/* If we're making a LFS, we break out here */
	r = make_lfs(fso, secsize, &dkw, minfree, bsize, fsize, lfs_segsize,
	    minfreeseg, resvseg, version, start, ibsize, interleave, roll_id);
	if (debug)
		bufstats();
	exit(r);
}

static int64_t
strsuftoi64(const char *desc, const char *arg, int64_t min, int64_t max, int *num_suffix)
{
	int64_t result, r1;
	int shift = 0;
	char	*ep;

	errno = 0;
	r1 = strtoll(arg, &ep, 10);
	if (ep[0] != '\0' && ep[1] != '\0')
		errx(1, "%s `%s' is not a valid number.", desc, arg);
	switch (ep[0]) {
	case '\0':
	case 's': case 'S':
		if (num_suffix != NULL)
			*num_suffix = 0;
		break;
	case 'g': case 'G':
		shift += 10;
		/* FALLTHROUGH */
	case 'm': case 'M':
		shift += 10;
		/* FALLTHROUGH */
	case 'k': case 'K':
		shift += 10;
		/* FALLTHROUGH */
	case 'b': case 'B':
		if (num_suffix != NULL)
			*num_suffix = 1;
		break;
	default:
		errx(1, "`%s' is not a valid suffix for %s.", ep, desc);
	}
	result = r1 << shift;
	if (errno == ERANGE || result >> shift != r1)
		errx(1, "%s `%s' is too large to convert.", desc, arg);
	if (result < min)
		errx(1, "%s `%s' (%" PRId64 ") is less than the minimum (%" PRId64 ").",
		    desc, arg, result, min);
	if (result > max)
		errx(1, "%s `%s' (%" PRId64 ") is greater than the maximum (%" PRId64 ").",
		    desc, arg, result, max);
	return result;
}

void
usage()
{
	fprintf(stderr, "usage: newfs_lfs [ -fsoptions ] special-device\n");
	fprintf(stderr, "where fsoptions are:\n");
	fprintf(stderr, "\t-A (autoconfigure segment size)\n");
	fprintf(stderr, "\t-B segment size in bytes\n");
	fprintf(stderr, "\t-D (debug)\n");
	fprintf(stderr, "\t-M count of segments not counted in bfree\n");
	fprintf(stderr,
	    "\t-N (do not create file system, just print out parameters)\n");
	fprintf(stderr, "\t-O first segment offset in sectors\n");
	fprintf(stderr, "\t-R count of segments reserved for the cleaner\n");
	fprintf(stderr, "\t-b block size in bytes\n");
	fprintf(stderr, "\t-f frag size in bytes\n");
	fprintf(stderr, "\t-m minimum free space %%\n");
	fprintf(stderr, "\t-s file system size in sectors\n");
	fprintf(stderr, "\t-v version\n");
	exit(1);
}
